types/state/stateTxn.go

```diff
@@ -0,0 +1,334 @@
+	}
+	tx.state.logInfo("AddTransaction", "chainId", logging.AsHex(chainId), "txid", logging.AsHex(txId), "pending", logging.AsHex(txPending.Entry), "accepted", logging.AsHex(txAcceptedEntry))
+
+	chainType, _ := binary.Uvarint(txPending.Entry)
+	if types.ChainType(chainType) != types.ChainTypePendingTransaction {
+		return fmt.Errorf("expecting pending transaction chain type of %s, but received %s",
+			types.ChainTypePendingTransaction.Name(), types.TxType(chainType).Name())
+	}
+
+	if txAccepted != nil {
+		chainType, _ = binary.Uvarint(txAccepted.Entry)
+		if types.ChainType(chainType) != types.ChainTypeTransaction {
+			return fmt.Errorf("expecting pending transaction chain type of %s, but received %s",
+				types.ChainTypeTransaction.Name(), types.ChainType(chainType).Name())
+		}
+	}
+
+	//append the list of pending Tx's, txId's, and validated Tx's.
+	tx.state.mutex.Lock()
+	defer tx.state.mutex.Unlock()
+
+	tsi := transactionStateInfo{txPending, chainId.Bytes(), txId}
+	tx.state.transactions.pendingTx = append(tx.state.transactions.pendingTx, &tsi)
+
+	if txAccepted != nil {
+		tsi := transactionStateInfo{txAccepted, chainId.Bytes(), txId}
+		tx.state.transactions.validatedTx = append(tx.state.transactions.validatedTx, &tsi)
+	}
+	return nil
+}
+
+// GetCurrentEntry retrieves the current state object from the database based upon chainId.  Current state either comes
+// from a previously saves state for the current block, or it is from the database
+func (tx *DBTransaction) GetCurrentEntry(chainId []byte) (*Object, error) {
+	if chainId == nil {
+		return nil, fmt.Errorf("chain id is invalid, thus unable to retrieve current entry")
+	}
+	var ret *Object
+	var err error
+	var key types.Bytes32
+
+	copy(key[:32], chainId[:32])
+
+	tx.state.mutex.Lock()
+	currentState := tx.state.updates[key]
+	tx.state.mutex.Unlock()
+	if currentState != nil {
+		ret = currentState.stateData
+	} else {
+		currentState := blockUpdates{}
+		currentState.bucket = bucketEntry
+		//pull current state entry from the database.
+		currentState.stateData, err = tx.GetPersistentEntry(chainId, false)
+		if err != nil {
+			return nil, err
+		}
+		//if we have valid data, store off the state
+		ret = currentState.stateData
+	}
+
+	return ret, nil
+}
+
+//AddSynthTx add the synthetic transaction which is mapped to the parent transaction
+func (tx *DBTransaction) AddSynthTx(parentTxId types.Bytes, synthTxId types.Bytes, synthTxObject *Object) {
+	tx.state.logInfo("AddSynthTx", "txid", logging.AsHex(synthTxId), "entry", logging.AsHex(synthTxObject.Entry))
+	var val *[]transactionStateInfo
+	var ok bool
+
+	parentHash := parentTxId.AsBytes32()
+	if val, ok = tx.state.transactions.synthTxMap[parentHash]; !ok {
+		val = new([]transactionStateInfo)
+		tx.state.transactions.synthTxMap[parentHash] = val
+	}
+	*val = append(*val, transactionStateInfo{synthTxObject, nil, synthTxId})
+}
+
+func (tx *DBTransaction) Commit(blockHeight int64) ([]byte, int, error) {
```

Comment:
Instead of adding a copy of `Commit`, rename `WriteStates` and change its receiver.